File: httpclient/src/main/java/com/gokhanaliccii/httpclient/JsonRequestQueue.java
package com.gokhanaliccii.httpclient;

import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.net.*;
import java.util.Map;

import static java.net.HttpURLConnection.HTTP_ACCEPTED;
import static java.net.HttpURLConnection.HTTP_OK;

public class JsonRequestQueue implements HttpRequestQueue {

    @Override
    public <T> void add(@NotNull HttpRequest<T> request, boolean needCache) {
        HttpRequestInfo requestInfo = request.getRequestInfo();
        HttpResult<T> resultListener = request.getResultListener();

    }

    @Override
    public <T> void add(@NotNull HttpRequest<T> request, boolean needCache, boolean shouldUseRetry) {
        HttpRequestInfo requestInfo = request.getRequestInfo();

        try {
            HttpURLConnection connection = (HttpURLConnection) new URL(requestInfo.getUrl()).openConnection();

            putHeaders(requestInfo, connection);
            putHttpMethod(requestInfo, connection);

            int responseCode = connection.getResponseCode();
            InputStream inputStream;
            if (HTTP_OK <= responseCode && HTTP_ACCEPTED <= 299) {
                inputStream = connection.getInputStream();
            } else {
                inputStream = connection.getErrorStream();
            }

            BufferedReader in = new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
            StringBuilder response = new StringBuilder();

            String currentLine;
            while ((currentLine = in.readLine()) != null)
                response.append(currentLine);

            in.close();

        } catch (Exception e) {
            e.printStackTrace();
        }


    }

    private void putHttpMethod(HttpRequestInfo requestInfo, HttpURLConnection connection) throws ProtocolException {
        connection.setRequestMethod(requestInfo.getMethod().name());
    }

    private void putHeaders(HttpRequestInfo requestInfo, HttpURLConnection connection) {
        Map<String, String> headers = requestInfo.getHeaders();
        for (String key : headers.keySet())
            connection.setRequestProperty(key, headers.get(key));
    }

    @Override
    public void remove(@NotNull String tag) {

    }

    @Override
    public void removeAll() {

    }

    @Override
    public void start() {

    }

    @Override
    public void stop() {

    }
}

File: httpclient/src/main/java/com/gokhanaliccii/httpclient/JsonRequestQueue.kt
package com.gokhanaliccii.httpclient

import android.os.Looper
import android.util.Log
import com.gokhanaliccii.httpclient.util.closeStreams
import com.gokhanaliccii.httpclient.util.isResponseCodeSucceed
import com.gokhanaliccii.jsonparser.jsonTo
import com.gokhanaliccii.jsonparser.jsonToList
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.ProtocolException
import java.net.URL
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class JsonRequestQueue : HttpRequestQueue {

    private val executorService: ExecutorService by lazy { Executors.newFixedThreadPool(10) }

    override fun <T> add(request: HttpRequestQueue.HttpRequest<T>, needCache: Boolean) {
    }

    override fun <T> add(request: HttpRequestQueue.HttpRequest<T>, needCache: Boolean, shouldUseRetry: Boolean) {
        sendRequest(request)
    }

    override fun start() {
    }

    override fun stop() {
    }

    override fun removeAll() {
    }

    private fun <T> sendRequest(request: HttpRequestQueue.HttpRequest<T>) {
        val requestInfo = request.requestInfo

        executorService.submit {
            try {
                val connection = URL(requestInfo.url).openConnection() as HttpURLConnection
                putHeaders(requestInfo, connection)
                putHttpMethod(requestInfo, connection)

                if (connection.responseCode.isResponseCodeSucceed()) {
                    val outputStream = ByteArrayOutputStream()
                    val inputStream = connection.inputStream
                    inputStream.copyTo(outputStream)
                    closeStreams(inputStream, outputStream)

                    val rawJson = outputStream.toString()
                    val returnType = requestInfo.returnType!!
                    val isReturnTypeArray = requestInfo.isArray!!

                    if (isReturnTypeArray) {
                        rawJson.jsonToList(returnType).let {
                            android.os.Handler(Looper.getMainLooper()).post {
                                request.resultListener.onResponse(it)
                            }
                        }
                    } else {
                        rawJson.jsonTo(returnType).let {
                            android.os.Handler(Looper.getMainLooper()).post {
                                request.resultListener.onResponse(it)
                            }
                        }
                    }

                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun putHeaders(requestInfo: HttpRequestQueue.HttpRequestInfo<*>, connection: HttpURLConnection) {
        val headers = requestInfo.headers
        for (key in headers.keys)
            connection.setRequestProperty(key, headers[key])
    }

    @Throws(ProtocolException::class)
    private fun putHttpMethod(requestInfo: HttpRequestQueue.HttpRequestInfo<*>, connection: HttpURLConnection) {
        connection.requestMethod = requestInfo.method.name
    }
}